File: mod10/Main.kt
fun findLongestChainEndpoint(pairs: Array<Pair<String, String>>): String {
    if (pairs.isEmpty()) {
        return "Нет пар для анализа"
    }

    val graph = mutableMapOf<String, String>()
    for (pair in pairs) {
        if (pair.first == pair.second) {
            return "обратитесь к специалисту"
        }
        graph[pair.first] = pair.second
    }

    var longestChainEndpoint = ""
    var maxLength = 0

    for (start in graph.keys) {
        var current = start
        var length = 1
        while (graph.containsKey(current)) {
            current = graph[current]!!
            length++
        }

        if (length > maxLength) {
            maxLength = length
            longestChainEndpoint = current
        }
    }

    return longestChainEndpoint
}

fun main() {
    println("Введите количество пар:")
    val numPairs = readLine()?.toIntOrNull() ?: 0

    if (numPairs <= 0) {
        println("Некорректный ввод количества пар.")
        return
    }

    val pairs = Array(numPairs) {
        println("Введите пару (начальный пункт конечный пункт):")
        val input = readLine()?.split(" ") ?: listOf() // Изменено здесь
        if (input.size == 2) {
            Pair(input[0], input[1])
        } else {
            println("Некорректный ввод пары. Попробуйте снова.")
            return@Array Pair("", "") // Возвращаем пустую пару и просим ввести снова
        }
    }

    //Проверка на пустые пары внутри массива.
    val filteredPairs = pairs.filter { it.first.isNotEmpty() && it.second.isNotEmpty() }.toTypedArray()

    val result = findLongestChainEndpoint(filteredPairs)
    println("Итоговый конечный пункт: $result")
}

File: mod10/main.cpp
#include <iostream>
#include <vector>
#include <map>
#include <string>

using namespace std;

string findLongestChainEndpoint(const vector<pair<string, string>>& pairs) {
    if (pairs.empty()) {
        return "Нет пар для анализа";
    }

    map<string, string> graph;
    for (const auto& pair : pairs) {
        if (pair.first == pair.second) {
            return "обратитесь к специалисту";
        }
        graph[pair.first] = pair.second;
    }

    string longestChainEndpoint = "";
    int maxLength = 0;

    for (const auto& start : graph) {
        string current = start.first;
        int length = 1;
        while (graph.count(current)) {
            current = graph[current];
            length++;
        }

        if (length > maxLength) {
            maxLength = length;
            longestChainEndpoint = current;
        }
    }

    return longestChainEndpoint;
}

int main() {
    cout << "Введите количество пар:" << endl;
    int numPairs;
    cin >> numPairs;

    if (numPairs <= 0) {
        cout << "Некорректный ввод количества пар." << endl;
        return 0;
    }

    vector<pair<string, string>> pairs(numPairs);
    cin.ignore(); // Очистить буфер после ввода числа

    for (int i = 0; i < numPairs; ++i) {
        cout << "Введите пару (начальный пункт конечный пункт):" << endl;
        string input;
        getline(cin, input);

        size_t spacePos = input.find(' ');
        if (spacePos != string::npos) {
            pairs[i] = make_pair(input.substr(0, spacePos), input.substr(spacePos + 1));
        } else {
            cout << "Некорректный ввод пары. Попробуйте снова." << endl;
            return 0;
        }
    }

    // Проверка на пустые пары внутри массива.
    vector<pair<string, string>> filteredPairs;
    for (const auto& pair : pairs) {
        if (!pair.first.empty() && !pair.second.empty()) {
            filteredPairs.push_back(pair);
        }
    }

    string result = findLongestChainEndpoint(filteredPairs);
    cout << "Итоговый конечный пункт: " << result << endl;

    return 0;
}